File: src/scripts/sim/citizenManager.js
import config from '../config.js';
import {CitizenState} from "../model/citizen.js";
import {findTile} from "./tileTools.js";
import {BuildingType} from "../model/buildings/buildingType.js";

export class CitizenManager {

    /**
     *
     * @param {World} world
     * @param {Citizen} citizen
     */
    static simulate(world, citizen) {
        if (citizen.residence === null) {
            if (citizen.workplace) {
                citizen.workplace.workers = citizen.workplace.workers.filter(
                    c => c.id !== citizen.id
                );
                citizen.workplace = null;

            }
            return false;
        }
        switch (citizen.state) {
            case CitizenState.idle:
            case CitizenState.school:
            case CitizenState.retired:
                // Action - None

                // Transitions - None

                break;
            case CitizenState.unemployed:
                // Action - Look for a job
                this.#findJob(world, citizen);

                // Transitions
                if (citizen.workplace) {
                    citizen.state = CitizenState.employed;
                    citizen.updated = true;
                }

                break;
            case CitizenState.employed:
                // Actions - None

                // Transitions
                if (!citizen.workplace) {
                    citizen.state = CitizenState.unemployed;
                    citizen.updated = true;
                }

                break;
            default:
                console.error(`Citizen ${citizen.id} is in an unknown state (${citizen.state})`);
        }
        return true;
    }

    /**
     *
     * @param {World} world
     * @param {Citizen} citizen
     */
    static #findJob(world, citizen) {
        const tile = findTile(world, citizen.residence.tile.x, citizen.residence.tile.y,
            (tile) => {
                // Search for an industrial or commercial building with at least one available job
                if (tile.building?.type === BuildingType.industrial ||
                    tile.building?.type === BuildingType.commercial) {
                    if (tile.building.availableJobs > 0) {
                        return true;
                    }
                }

                return false;
            }, config.citizen.maxJobSearchDistance);
        // TODO: feels like this should happen in `simulate()`
        if (tile) {
            // Employ the citizen at the building
            tile.building.workers.push(citizen);
            citizen.workplace = tile.building;
        }
    }
}

File: Simulation/Systems/CitizenManager.cs
using Simulation.Model.Buildings;
using Simulation.Model.Buildings.Zones;

namespace Simulation.Systems;

using Model;

public static class CitizenManager
{
    /// <summary>
    /// Advances a citizen's state machine by one tick.
    /// Returns false if the citizen should be removed from the world (no residence).
    /// </summary>
    public static bool Simulate(World world, Citizen citizen)
    {
        if (citizen.Residence is null)
        {
            // Citizen lost their home — sever workplace link and remove from world.
            if (citizen.Workplace is { } workplace)
            {
                workplace.Workers.Remove(citizen);
                workplace.Updated = workplace.Tile.Updated = true;
                citizen.Workplace = null;
            }
            return false;
        }

        switch (citizen.State)
        {
            case CitizenState.Idle:
            case CitizenState.School:
            case CitizenState.Retired:
                break;

            case CitizenState.Unemployed:
                FindJob(world, citizen);
                if (citizen.Workplace is not null)
                {
                    citizen.State = CitizenState.Employed;
                    citizen.Residence.Updated = citizen.Residence.Tile.Updated = true;
                }
                break;

            case CitizenState.Employed:
                if (citizen.Workplace is null)
                {
                    citizen.State = CitizenState.Unemployed;
                    citizen.Residence.Updated = citizen.Residence.Tile.Updated = true;
                }
                break;
        }
        return true;
    }

    private static void FindJob(World world, Citizen citizen)
    {
        var tile = TileTools.FindTile(
            world,
            citizen.Residence!.Tile.X,
            citizen.Residence.Tile.Y,
            t => t.Building is ZoneWithJobs
            {
                Type: BuildingType.Commercial or BuildingType.Industrial, 
                AvailableJobs: > 0
            },
            SimConfig.Citizen.MaxJobSearchDistance);

        if (tile?.Building is ZoneWithJobs workplace)
        {
            workplace.Workers.Add(citizen);
            citizen.Workplace    = workplace;
            workplace.Updated    = tile.Updated = true;
        }
    }
}
